protocols/gossipsub/src/behaviour.rs

```diff
@@ -0,0 +1,1176 @@
+                "GRAFT: Not subscribed to topics -  Sending PRUNE to peer: {:?}",
+                peer_id
+            );
+            self.events.push_back(NetworkBehaviourAction::SendEvent {
+                peer_id: peer_id.clone(),
+                event: Arc::new(GossipsubRpc {
+                    subscriptions: Vec::new(),
+                    messages: Vec::new(),
+                    control_msgs: prune_messages,
+                }),
+            });
+        }
+        debug!("Completed GRAFT handling for peer: {:?}", peer_id);
+    }
+
+    /// Handles PRUNE control messages. Removes peer from the mesh.
+    fn handle_prune(&mut self, peer_id: &PeerId, topics: Vec<TopicHash>) {
+        debug!("Handling PRUNE message for peer: {:?}", peer_id);
+        for topic_hash in topics {
+            if let Some(peers) = self.mesh.get_mut(&topic_hash) {
+                // remove the peer if it exists in the mesh
+                info!(
+                    "PRUNE: Removing peer: {:?} from the mesh for topic: {:?}",
+                    peer_id, topic_hash
+                );
+                peers.retain(|p| p != peer_id);
+                //TODO: untagPeer
+            }
+        }
+        debug!("Completed PRUNE handling for peer: {:?}", peer_id);
+    }
+
+    /// Handles a newly received GossipsubMessage.
+    /// Forwards the message to all peers in the mesh.
+    fn handle_received_message(&mut self, msg: GossipsubMessage, propagation_source: &PeerId) {
+        debug!(
+            "Handling message: {:?} from peer: {:?}",
+            msg.id(),
+            propagation_source
+        );
+        // if we have seen this message, ignore it
+        // there's a 3% chance this is a false positive
+        // TODO: Check this has no significant emergent behaviour
+        if self.received.put(msg.id(), ()).is_some() {
+            info!(
+                "Message already received, ignoring. Message: {:?}",
+                msg.id()
+            );
+            return;
+        }
+
+        // add to the memcache
+        self.mcache.put(msg.clone());
+
+        // dispatch the message to the user
+        if self.mesh.keys().any(|t| msg.topics.iter().any(|u| t == u)) {
+            debug!("Sending received message to user");
+            self.events.push_back(NetworkBehaviourAction::GenerateEvent(
+                GossipsubEvent::Message(propagation_source.clone(), msg.clone()),
+            ));
+        }
+
+        // forward the message to mesh peers, if no validation is required
+        if !self.config.manual_propagation {
+            let message_id = msg.id();
+            self.forward_msg(msg, propagation_source);
+            debug!("Completed message handling for message: {:?}", message_id);
+        }
+    }
+
+    /// Handles received subscriptions.
+    fn handle_received_subscriptions(
+        &mut self,
+        subscriptions: &[GossipsubSubscription],
+        propagation_source: &PeerId,
+    ) {
+        trace!(
```

Comment:
```suggestion
        debug!(
```

This is on `debug` for all the other functions. Is this one to noisy?